spec.md

```diff
@@ -1090,6 +1086,220 @@ message ControllerServiceCapability {
+
+  // Secrets required by plugin to complete snapshot creation request.
+  // This field is OPTIONAL.
+  map<string, string> create_snapshot_secrets = 3;
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  // This field is OPTIONAL. The Plugin is responsible for parsing and
+  // validating these parameters. COs will treat these as opaque.
+  // Use cases for opaque parameters:
+  // - Specify a policy to automatically clean up the snapshot.
+  // - Specify an expiration date for the snapshot.
+  // - Specify whether the snapshot is readonly or read/write.
+  // - Specify if the snapshot should be replicated to some place.
+  // - Specify primary or secondary for replication systems that
+  //   support snapshotting only on primary.
+  map<string, string> parameters = 4;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
+  // This is the complete size of the snapshot in bytes. The purpose of
+  // this field is to give CO guidance on how much space is needed to
+  // create a volume from this snapshot. The size of the volume must not
+  // be less than the size of the source snapshot. This field is
+  // OPTIONAL. If this field is not set, it indicates that this size is
+  // unknown. The value of this field MUST NOT be negative.
+  SnapshotSize size = 1;
+
+  // Uniquely identifies a snapshot and is generated by the plugin. It
+  // will not change over time. This field is REQUIRED. The identity
+  // information will be used by the CO in subsequent calls to refer to
+  // the provisioned snapshot.
+  string id = 2;
+
+  // Identity information for the source volume. Note that create
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string source_volume_id = 3;
+
+  // Timestamp when the point-in-time snapshot is taken on the storage
+  // system. The format of this field should be a Unix Nanoseconds time
+  // encoded as an int64. This field is REQUIRED.
+  int64 created_at = 4;
+
+  // The status of a snapshot.
+  SnapshotStatus status = 5;
+}
+
+message SnapshotSize {
+  // The complete size of the snapshot in bytes.
+  int64 size_bytes = 1;
+}
+
+// The status of a snapshot.
+message SnapshotStatus {
+  enum Type {
+     UNKNOWN = 0;
+     // A snapshot is ready for use.
+     READY = 1;
+     // A snapshot is cut and is now being uploaded.
+     // Some cloud providers and storage systems uploads the snapshot
+     // to the cloud after the snapshot is cut. During this phase,
+     // `thaw` can be done so the application can be running again if
+     // `freeze` was done before taking the snapshot.
+     UPLOADING = 2;
+     // A snapshot is in error status.
```

Comment:
is this a terminal status? as in, the snapshot can only be deleted from here?

Reply:
If a snapshot ends up in error status, the user will need to find out what has caused the error condition and then determines what to do with it.  If an error happens before the snapshot is cut, it needs to be deleted.  If an error happens during the upload, may be the cloud admin has a way to fix the problem and still allow the snapshot to be used.

Reply:
is the user the only actor that can initiate recovery here? for example, are there ways that a CO could gracefully recovery from an ERROR condition and if so, what are they? (read: please document them if there are any)

Reply:
I checked with Jing on the upload case.  If an error happens during upload, that is also a terminal status so the snapshot can be deleted.  So that is consistent with the error before the snapshot is cut.